In [1]:
import pandas as pd
bridgesdf = pd.read_csv("bridges.csv")

The column of interest in this dataset is length. Assuming that this column is measured in feet, a plausible range would be between ~0.2 miles (1000 feet) and ~0.75 mile (4000 feet). The reasoning for the lower bound is that, if a waterway is sufficiently narrow, the land surrounding it is often redeveloped to avert the flow entirely out of a highly-trafficked area. As for the upper bound, bridges usually are not a full mile long because such a large building project can be more costly than exploring an alternative, such as providing ferry service across a wide body of water. The dataset will be screened in accordance with this assumption.<p>

"length" lower bound: 1000 <br>
"length" upper bound: 4000

In [2]:
bridgesdf.shape

(108, 13)

In [3]:
bridgesdf.head(10)

,Id,river,location,erected,purpose,length,lanes,clear-g,t-or-d,material,span,rel-l,type
0,E1,M,3.0,1818,HIGHWAY,NaN,2.0,N,THROUGH,WOOD,SHORT,S,WOOD
1,E2,A,25.0,1819,HIGHWAY,1037.0,2.0,N,THROUGH,WOOD,SHORT,S,WOOD
2,E3,A,39.0,1829,AQUEDUCT,NaN,1.0,N,THROUGH,WOOD,NaN,S,WOOD
3,E5,A,29.0,1837,HIGHWAY,1000.0,2.0,N,THROUGH,WOOD,SHORT,S,WOOD
4,E6,M,23.0,1838,HIGHWAY,NaN,2.0,N,THROUGH,WOOD,NaN,S,WOOD
5,E7,A,27.0,1840,HIGHWAY,990.0,2.0,N,THROUGH,WOOD,MEDIUM,S,WOOD
6,E8,A,28.0,1844,AQUEDUCT,1000.0,1.0,N,THROUGH,IRON,SHORT,S,SUSPEN
7,E9,M,3.0,1846,HIGHWAY,1500.0,2.0,N,THROUGH,IRON,SHORT,S,SUSPEN
8,E10,A,39.0,1848,AQUEDUCT,NaN,1.0,N,DECK,WOOD,NaN,S,WOOD
9,E11,A,29.0,1851,HIGHWAY,1000.0,2.0,N,THROUGH,WOOD,MEDIUM,S,WOOD


In [4]:
bridgesdf.info()

<class 'pandas.DataFrame'>
RangeIndex: 108 entries, 0 to 107
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Id        108 non-null    str    
 1   river     108 non-null    str    
 2   location  107 non-null    float64
 3   erected   108 non-null    int64  
 4   purpose   108 non-null    str    
 5   length    81 non-null     float64
 6   lanes     92 non-null     float64
 7   clear-g   106 non-null    str    
 8   t-or-d    102 non-null    str    
 9   material  106 non-null    str    
 10  span      92 non-null     str    
 11  rel-l     103 non-null    str    
 12  type      106 non-null    str    
dtypes: float64(3), int64(1), str(9)
memory usage: 11.1 KB


In [5]:
round(bridgesdf.describe(),4)

,location,erected,length,lanes
count,107.0000,108.0000,81.0000,92.0000
mean,25.9785,1905.3148,1567.4691,2.6304
std,13.6659,37.1737,747.4915,1.1647
min,1.0000,1818.0000,804.0000,1.0000
25%,15.5000,1884.0000,1000.0000,2.0000
50%,27.0000,1903.0000,1300.0000,2.0000
75%,37.5000,1928.0000,2000.0000,4.0000
max,52.0000,1986.0000,4558.0000,6.0000


We can see that there are datapoints that lie outside our assumed bounds. Let us identify the outliers for "length" so that they can be trimmed out of the dataset.

In [6]:
bridgesdf [ (bridgesdf ["length"] < 1000) | (bridgesdf ["length"] > 4000) ]

,Id,river,location,erected,purpose,length,lanes,clear-g,t-or-d,material,span,rel-l,type
5,E7,A,27.0,1840,HIGHWAY,990.0,2.0,N,THROUGH,WOOD,MEDIUM,S,WOOD
31,E34,O,41.0,1888,RR,4558.0,2.0,G,THROUGH,STEEL,LONG,F,SIMPLE-T
56,E53,A,28.0,1904,RR,965.0,4.0,G,THROUGH,STEEL,MEDIUM,S-F,SIMPLE-T
70,E64,A,29.0,1923,HIGHWAY,885.0,4.0,G,THROUGH,STEEL,MEDIUM,F,ARCH
72,E70,A,27.0,1926,HIGHWAY,860.0,4.0,G,THROUGH,STEEL,MEDIUM,S-F,SUSPEN
73,E69,A,26.0,1926,HIGHWAY,884.0,4.0,G,THROUGH,STEEL,MEDIUM,S-F,SUSPEN
80,E71,A,25.0,1928,HIGHWAY,860.0,4.0,G,THROUGH,STEEL,MEDIUM,S-F,SUSPEN
88,E107N,A,39.7,1945,RR,840.0,2.0,G,THROUGH,STEEL,MEDIUM,S-F,SIMPLE-T
94,E98,M,22.0,1951,HIGHWAY,900.0,4.0,G,THROUGH,STEEL,MEDIUM,F,CONT-T
98,E82,O,42.0,1955,HIGHWAY,804.0,NaN,G,THROUGH,STEEL,NaN,F,SIMPLE-T


We can see that exactly one "length" outlier is above 4000. The other 12 returned rows are below 1000, but some are only barely below 1000, such as "Id"==E7, which has "length"==990. The rows with "length"<1000 are distributed evenly and none of them are excessively below "length"==1000, so they represent genuine and relevant data points. For this reason, we will not consider them outliers, as excluding them from the data would artificially skew any subsequent analysis.

<p>We will only screen the "length"==4558 outlier out of the data by dropping that row from the dataset, as it is an isolated value far above the rest.

In [7]:
bridgesdf.drop([31])

,Id,river,location,erected,purpose,length,lanes,clear-g,t-or-d,material,span,rel-l,type
0,E1,M,3.0,1818,HIGHWAY,NaN,2.0,N,THROUGH,WOOD,SHORT,S,WOOD
1,E2,A,25.0,1819,HIGHWAY,1037.0,2.0,N,THROUGH,WOOD,SHORT,S,WOOD
2,E3,A,39.0,1829,AQUEDUCT,NaN,1.0,N,THROUGH,WOOD,NaN,S,WOOD
3,E5,A,29.0,1837,HIGHWAY,1000.0,2.0,N,THROUGH,WOOD,SHORT,S,WOOD
4,E6,M,23.0,1838,HIGHWAY,NaN,2.0,N,THROUGH,WOOD,NaN,S,WOOD
...,...,...,...,...,...,...,...,...,...,...,...,...,...
103,E84,A,24.0,1969,HIGHWAY,870.0,6.0,G,THROUGH,STEEL,MEDIUM,F,ARCH
104,E91,O,44.0,1975,HIGHWAY,3756.0,6.0,G,THROUGH,STEEL,LONG,F,ARCH
105,E90,M,7.0,1978,HIGHWAY,950.0,6.0,G,THROUGH,STEEL,LONG,F,ARCH
106,E100,O,43.0,1982,HIGHWAY,NaN,NaN,G,NaN,NaN,NaN,F,NaN


In [8]:
bridgesdf = bridgesdf.drop([31])

In [9]:
bridgesdf.shape

(107, 13)

In [10]:
bridgesdf [ (bridgesdf ["length"] > 4000) ]

,Id,river,location,erected,purpose,length,lanes,clear-g,t-or-d,material,span,rel-l,type


The outlier with "length">4000 has been dropped from the dataset. Let's create a new csv so that the cleaned data is ready for use.

In [11]:
bridgesdf.to_csv("bridges_cleaned.csv")

I consulted the pandas User Guide to write this lab:<p>
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html#pandas.DataFrame.drop

Github Repo URL: https://github.com/journeyrahman/cs82a-portfolio